apps/server/src/modules/tldraw/service/tldraw.ws.service.spec.ts

```diff
@@ -0,0 +1,376 @@
+import { encoding } from 'lib0';
+import { TldrawBoardRepo } from '@src/modules/tldraw/repo';
+import { TldrawWs } from '@src/modules/tldraw/controller';
+import { TldrawWsService } from '.';
+import { TestHelper } from '../helper/test-helper';
+
+describe('TldrawWS', () => {
+	let app: INestApplication;
+	let ws: WebSocket;
+	let service: TldrawWsService;
+
+	const gatewayPort = 3346;
+	const wsUrl = TestHelper.getWsUrl(gatewayPort);
+	const testMessage =
+		'AZQBAaCbuLANBIsBeyJpZCI6ImU1YTYwZGVjLTJkMzktNDAxZS0xMDVhLWIwMmM0N2JkYjFhMiIsInRkVXNlciI6eyJp' +
+		'ZCI6ImU1YTYwZGVjLTJkMzktNDAxZS0xMDVhLWIwMmM0N2JkYjFhMiIsImNvbG9yIjoiI0JENTRDNiIsInBvaW50Ijpb' +
+		'ODY4LDc2OV0sInNlbGVjdGVkSWRzIjpbXSwiYWN0aXZlU2hhcGVzIjpbXSwic2Vzc2lvbiI6ZmFsc2V9fQ==';
+
+	const delay = (ms: number) =>
+		new Promise((resolve) => {
+			setTimeout(resolve, ms);
+		});
+
+	jest.useFakeTimers();
+
+	beforeAll(async () => {
+		const imports = [CoreModule, ConfigModule.forRoot(createConfigModuleOptions(config))];
+		const testingModule = await Test.createTestingModule({
+			imports,
+			providers: [TldrawWs, TldrawBoardRepo, TldrawWsService],
+		}).compile();
+
+		service = testingModule.get<TldrawWsService>(TldrawWsService);
+		app = testingModule.createNestApplication();
+		app.useWebSocketAdapter(new WsAdapter(app));
+		jest.useFakeTimers({ advanceTimers: true, doNotFake: ['setInterval', 'clearInterval', 'setTimeout'] });
+		await app.init();
+	});
+
+	afterAll(async () => {
+		await app.close();
+	});
+
+	const createMessage = (values: number[]) => {
+		const encoder = encoding.createEncoder();
+		values.forEach((val) => {
+			encoding.writeVarUint(encoder, val);
+		});
+		encoding.writeVarUint(encoder, 0);
+		encoding.writeVarUint(encoder, 1);
+		const msg = encoding.toUint8Array(encoder);
+		return {
+			msg,
+		};
+	};
+
+	it('should service properties be defined', () => {
+		expect(service).toBeDefined();
+		expect(service.pingTimeout).toBeDefined();
+		expect(service.persistence).toBeDefined();
+	});
+
+	describe('when client is not connected to WS', () => {
+		const setup = async () => {
+			ws = await TestHelper.setupWs(wsUrl);
+
+			const closeConSpy = jest.spyOn(service, 'closeConn').mockImplementationOnce(() => {});
+			const sendSpy = jest.spyOn(service, 'send');
+			const doc: { conns: Map<WebSocket, Set<number>> } = { conns: new Map() };
+			const byteArray = new TextEncoder().encode(testMessage);
+
+			return {
+				closeConSpy,
+				sendSpy,
+				doc,
+				byteArray,
+			};
+		};
+
+		it('should throw error for send message', async () => {
+			const { closeConSpy, sendSpy, doc, byteArray } = await setup();
+
+			service.send(doc as WsSharedDocDo, ws, byteArray);
+
+			expect(sendSpy).toThrow();
+			expect(sendSpy).toHaveBeenCalledWith(doc, ws, byteArray);
+			expect(closeConSpy).toHaveBeenCalled();
+
+			ws.close();
+			sendSpy.mockRestore();
+		});
+	});
+
+	describe('when websocket has ready state different than 0 or 1', () => {
```

Comment:
where do I find the tests for 0 and 1?

Reply:
Tests for 0 or 1 and not 0 or 1 are now beside each other